core/src/main/java/org/apache/calcite/rel/rules/FilterExtractInnerJoinRule.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Planner rule that matches an {@link org.apache.calcite.rel.core.Aggregate}
+ * on a {@link org.apache.calcite.rel.core.Join} and removes the left input
+ * of the join provided that the left input is also a left join if possible.
+ *
+ * <p>For instance,
+ *
+ * <blockquote>
+ * <pre>select distinct s.product_id, pc.product_id
+ * from sales as s
+ * left join product as p
+ *   on s.product_id = p.product_id
+ * left join product_class pc
+ *   on s.product_id = pc.product_id</pre></blockquote>
+ *
+ * <p>becomes
+ *
+ * <blockquote>
+ * <pre>select distinct s.product_id, pc.product_id
+ * from sales as s
+ * left join product_class pc
+ *   on s.product_id = pc.product_id</pre></blockquote>
+ *
+ * @see CoreRules#AGGREGATE_JOIN_JOIN_REMOVE
+ */
+public class FilterExtractInnerJoinRule
+    extends RelRule<FilterExtractInnerJoinRule.Config>
+    implements TransformationRule {
+
+  protected FilterExtractInnerJoinRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Filter filter = call.rel(0);
+    final Join join = call.rel(1);
+    RelBuilder builder = call.builder();
+
+    if (!isCrossJoin(join, builder)) {
+      return;
+    }
+
+    Stack<Map.Entry<RelNode, Integer>> stackForTablesWithEndIndexes = new Stack<>();
+    List<RexNode> allConditions = new ArrayList<>();
+    populateStackWithEndIndexesForTables(join, stackForTablesWithEndIndexes, allConditions);
+    if (((RexCall) filter.getCondition()).getOperands().stream().allMatch(operand ->
+        operand instanceof RexCall)) {
+      allConditions.addAll(((RexCall) filter.getCondition()).getOperands());
+    } else {
+      allConditions.add(filter.getCondition());
+    }
+
+    final RelNode modifiedJoinClauseWithWhereClause =
+        moveConditionsFromWhereClauseToJoinOnClause(
+            allConditions, stackForTablesWithEndIndexes, builder);
+
+    call.transformTo(modifiedJoinClauseWithWhereClause);
+  }
+
+  private static boolean isCrossJoin(Join join, RelBuilder builder) {
+    return join.getJoinType().equals(JoinRelType.INNER)
+        && builder.literal(true).equals(join.getCondition());
+  }
+  private void populateStackWithEndIndexesForTables(
+      Join join, Stack<Map.Entry<RelNode, Integer>> stack, List<RexNode> joinConditions) {
+    RelNode left = ((HepRelVertex) join.getLeft()).getCurrentRel();
+    RelNode right = ((HepRelVertex) join.getRight()).getCurrentRel();
+    int leftTableColumnSize = join.getLeft().getRowType().getFieldCount();
+    int rightTableColumnSize = join.getRight().getRowType().getFieldCount();
+    stack.push(Map.entry(right, leftTableColumnSize + rightTableColumnSize - 1));
+    if (!(join.getCondition() instanceof RexLiteral)) {
+      joinConditions.add(join.getCondition());
+    }
+    if (left instanceof Join) {
+      populateStackWithEndIndexesForTables((Join) left, stack, joinConditions);
+    } else {
+      stack.push(Map.entry(left, leftTableColumnSize - 1));
```

Comment:
Check if creating new Entry object using Map.Entry syntax works with the Java version configured for Calcite

Reply:
Map.Entry is working with Java 11 but it's not supported in Java 8.